docs: tutorial for bindings with Kafka

CynthiaPeter opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

This is an initial draft for a binding with Kafka tutorial.

CynthiaPeter committed:
Initial draft for bindings with Kafka tutorials

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+
+## Introduction
+
+In this tutorial, you will learn how bindings with Kafka works. You will learn how to extend your AsyncAPI document with Protocol-specific information, such as how to use Kafka bindings. You will also learn about server bindings and channel bindings.
```
```suggestion
In this tutorial, you will learn how bindings with Kafka works. You will learn how to extend your AsyncAPI document with Protocol-specific information, such as how to specify kafka topic specific configurations or how to defined kafka message key content.
```

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+Kafka is a messaging system that allows for the efficient handling and processing of data streams. 
+
+**What is binding?**
+In the context of Kafka, "bindings" refer to the connections between Kafka and other systems or applications. There are several levels of bindings as described below:
```
i'm not sure that this is the correct definition maybe 
```suggestion
In the context of Kafka, "bindings" refer to the way the kafka ecosystem is configured. There are several levels of bindings as described below:
```

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+**What is binding?**
+In the context of Kafka, "bindings" refer to the connections between Kafka and other systems or applications. There are several levels of bindings as described below:
+
+<!---How many levels or types of binding are there exactly and where is the best place to place them in this tutorial?--->
```
add a ref to https://github.com/asyncapi/bindings/tree/master/kafka

CynthiaPeter replied:
How many levels or types of bindings are there exactly? And where is the best place to place them in this tutorial? I considered putting the list under background context. However, I will need to show how these bindings work individually somewhere in the step-by-step section.

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+
+- Server Binding
+- Channel binding
+- Operation Binding
```
```suggestion
- Operation Bindings
- Message Bindings
```

CynthiaPeter replied:
Hello @M3lkior,

Should I spell all three of them as "Bindings," or its just Bindings for Operations?

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+Bindings: Bindings are relevant when integrating Kafka with other programs or services. A binding is a specific agreement or configuration that governs how an application connects to and interacts with Kafka. It can include information such as:
+
+- Which topics should an application read from or write to?
+- How should messages be formatted?
```
Those two parts are directly supported by asyncapi thanks to the specification

CynthiaPeter replied:
Does this mean that the third one is not supported?

M3lkior commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,66 @@
+- How should messages be formatted?
+- What are the rules for interacting with various forms of data or messages?
+
+Bindings are vital because they ensure that applications and Kafka can communicate effectively with each other's message formats, protocols, and expectations, ensuring data flows securely and efficiently within a system.
```
```suggestion
Bindings are importants because they show indications about how Kafka components are configured.
```

CynthiaPeter replied:
Thank you so much for taking the time to review this first draft. I appreciate the feedback and will have an updated version by the end of today.

CynthiaPeter commented:
Some questions I have on this task:

- Is there an installation guide needed for this topic?
- What can I build to ensure the reader understands and follows through?
- How best do I structure this?
- Where is the best place to let the reader know that they need an AsyncAPI file and maybe (Do they need to set up Kafka?)
- "Topics are split into partitions; each partition can contain a copy of the messages sent through Kafka. This can be important if you have bindings related to partitions". Jose shared this, and because he mentioned it, I'll add it as a footnote here and think through it. (I thought I'd add it to the Channel binding object sub-topic since that is where topics and partitions are defined as fields).

Answered
- Do I need to build an app to demonstrate this topic? - An AsyncAPI file? - Yes. I will be using Joy's AsyncAPI document.

CynthiaPeter committed:
Added feedback from Ludovic and moved the questions to the comments section

CynthiaPeter committed:
Added the first version of a diagram

derberg commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,83 @@
+---
+title: Bindings with Kafka
```
```suggestion
title: Adding Kafka information using bindings
```

maybe?

derberg commented on pages/docs/tutorials/bindings-with-kafka.md:
```diff
@@ -0,0 +1,83 @@
+
+Now, let’s see how to use bindings with Kafka.
+
+## Installation Guide
```
what installation you have in mind? bindings are extra properties one can add to AsyncAPI document.

quetzalliwrites replied:
No installation is needed in this tutorial, she already knows to update accordingly :)

derberg reviewed (commented):
as far as I remember this was suppose to be a tutorial, one of a series of Kafka tutorials

this one is a tutorial that will be visible after tutorial: `where you grab existing AsyncAPI Kafka document from previous tutorials and put schemas in Apicurio schema registry` (@Arya-Gupta works on this one afaik)

so in your tutorial you introduce bindings, and you do not have to introduce Kafka again (it should be already done in previous tutorials 🤔 ). So this tutorial explains how to add specific bindings to AsyncAPI document:
- server with info about schema registry
- channel or operation binding (depending on the initial shape of AsyncAPI document)

please also learn from https://github.com/asyncapi/website/pull/2042 what bindings are in AsyncAPI as `Bindings: Bindings are relevant when integrating Kafka with other programs or services. A binding is a specific agreement or configuration that governs how an application connects to and interacts with Kafka` sounds like bindings are some Kafka feature

CynthiaPeter committed:
Rewrote the document

VaishnaviNandakumar commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,38 @@
+
+## Introduction
+
+In this tutorial, you will learn how bindings with Kafka works. You will learn how to extend your AsyncAPI document with Protocol-specific information, such as how to specify Kafka topic-specific configurations or how to define Kafka message key content.
```
```suggestion
In this tutorial, you will learn how bindings with Kafka works. You will learn how to extend your AsyncAPI document with protocol-specific information, such as specifying configurations unique to Kafka topics or defining Kafka message key content.
```
The word specific/specify seems repetitive here.

quetzalliwrites committed:
Merge branch 'master' into feature/tutorial-bindings-with-kafka

quetzalliwrites commented on pages/docs/tutorials/kafka/index.md:
```diff
@@ -0,0 +1,4 @@
+---
+title: Tutorial-1
```
Why is the title "Tutorial-1"? Please fix this line with an appropriate title :)

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,38 @@
+
+- Which topics should an application read from or write to?
+- How should messages be formatted?
+- What are the rules for interacting with various data or messages?
```
- Let's consider turning this bullet point list into sentence format and include it in the previous paragraph from line 12.
- CONFIRM: Is this info `required` or optional? If optional, prob not needed.
**- Determine if these questions should be turned into statements.**

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,38 @@
+
+
+
+Now, let's see how to use bindings with Kafka.
```
```suggestion
```

Let's remove, not needed in this section :)

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,38 @@
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on two levels of bindings:
+
+- Server Bindings
+- Channel Bindings
```
```suggestion
You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on two levels of bindings: Server Bindings and Channel Bindings.
```

Add links to the object's documentation :)

CynthiaPeter committed:
implemented chnages and feedback

CynthiaPeter committed:
implemented feedback

CynthiaPeter committed:
Updated docs

CynthiaPeter committed:
Made changes based on feedback received

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,40 @@
+
+## Introduction
+
+In a previous tutorial, you learnt how to manage schemas with schema registry. In this tutorial, you will learn how bindings with Kafka works. You will learn how to define Kafka message key content and extend your AsyncAPI document with protocol-specific information.
```
```suggestion
In a previous tutorial, you learnt how to manage schemas with schema registry. In this tutorial, you will learn how bindings with Kafka works by defining Kafka message key content and extending your AsyncAPI document with protocol-specific information.
```

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,40 @@
+
+Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes to, how the messages are formatted, and the rules for interacting with various data or messages. 
+
+Bindings configure ecosystems to accommodate specific functionalities not supported by AsyncAPI's core features.
```
bindings configure nothing in the kafka ecosystems but allow documenting how the stack is configured
```suggestion
Bindings definitions allow you to specify the protocol specific functionalities not supported by AsyncAPI's core features.
```

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,40 @@
+Bindings configure ecosystems to accommodate specific functionalities not supported by AsyncAPI's core features.
+
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on two levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object) and [channel bindings](https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object). 
+
```
where to not cover all the bindings levels ? Kafka key bindings are more essential than server bindings for example; if you choose to keep only two levels; do a focus on channel + message key bindings

CynthiaPeter replied:
Thanks for this feedback. I will discuss this with Alejandra and prioritize properly.

quetzalliwrites replied:
Cool, so Cynthia is just gonna do 3 levels then: server, channel, and messages.

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,40 @@
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on two levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object) and [channel bindings](https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object). 
+
+
+The diagram below shows how bindings with Kafka work. 
```
i don't see any diagram :)

CynthiaPeter replied:
My bad. I had several versions, so I wanted to get your feedback on Slack.

quetzalliwrites replied:
**Diagram design style guide:** 
- use this blue #47BCEE for the highlighted component(s) in your diagram
- default mermaid diagram coloring for remaining nodes (rectangles)

CynthiaPeter committed:
Added a new version for diagram

CynthiaPeter committed:
Updated the title to make it brief

CynthiaPeter committed:
updated based on feedback

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,53 @@
+In a previous tutorial, you learnt how to manage schemas with schema registry. In this tutorial, you will learn how bindings with Kafka works by defining Kafka message key content and extending your AsyncAPI document with protocol-specific information.
+
+## Background Context
+
```
For the background context...
- Define the general concept of `bindings`
- Tell me it's relationship to the AsyncAPI document
- Define Kafka `bindings`

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,53 @@
+
+## Background Context
+
+Bindings enhances AsyncAPI documentation by providing protocol-specific information. They append to different document sections, such as servers, channels, or messages, and they include standard information that is unique to a particular protocol. 
```
```suggestion
Bindings provide protocol-specific information for your Event-Driven application. Protocols have specific ways of handling messages, and bindings abstract these details away from your application's core logic. Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes to, how the messages are formatted, and the rules for interacting with various data or messages. 

In an AsyncAPI document, bindings can be appended to different document sections, such as servers, channels, or messages. They include standard information that is unique to a particular protocol. Bindings definitions allow you to specify the protocol-specific functionalities not supported by AsyncAPI's core features.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,53 @@
+
+Bindings enhances AsyncAPI documentation by providing protocol-specific information. They append to different document sections, such as servers, channels, or messages, and they include standard information that is unique to a particular protocol. 
+
+Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes to, how the messages are formatted, and the rules for interacting with various data or messages. 
```
```suggestion
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,53 @@
+
+Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes to, how the messages are formatted, and the rules for interacting with various data or messages. 
+
+Bindings definitions allow you to specify the protocol specific functionalities not supported by AsyncAPI's core features.
```
```suggestion
```

Commit:
Apply suggestions from code review

CynthiaPeter committed:
Updated

CynthiaPeter committed:
updated based on feedback received

CynthiaPeter committed:
Updated background context

derberg commented on pages/docs/tutorials/kafka/index.md:
```diff
@@ -0,0 +1,4 @@
+---
```
do you need this?

CynthiaPeter replied:
I assume so. It shows the start and end of the metadata. And that's whats across the docs. Except I am missing something.

derberg replied:
sorry, I meant the file not `---`. I guess you have it just for the sake of PR preview?

derberg replied:
best check with other docs contributors for Kafka,
not sure maybe you agreed you need this file

CynthiaPeter replied:
The index.md and section.md files are default files for folders created for documentation.

quetzalliwrites replied:
Heyo @CynthiaPeter, since we will merge Joy's tutorials first because they are the start of our series, then your PRs do not need to include this `_section.md` or `index.md` files. You can remove these from your PR and keep just your tutorial. 

let me know if that makes sense 😸

CynthiaPeter replied:
Done.

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
````diff
@@ -0,0 +1,50 @@
+
+The diagram below shows how bindings with Kafka work. 
+
+```mermaid
````
I don't think we need such diagram that explains kafka in asyncapi

we already have a good doc and diagram in https://www.asyncapi.com/docs/concepts/asyncapi-document/adding-bindings

and your intro already links to kafka specific bindings

CynthiaPeter replied:
Ok. This means that we don't need an image here, right?

derberg replied:
diagrams are always nice, but only if they bring some value in given context. As I wrote in previous comment, there is already a nice binding overview diagram, so maybe link there. Just this diagram do not help much - I basically do not understand it, there is no text that describes it - so for now I advocate to remove it

CynthiaPeter replied:
Ok. Thank you.

derberg reviewed (commented):
lgmt so far, great and precise job 👏🏼 

sorry for keeping you waiting 😞 

just remember that when you will be adding examples of bindings for each section, this tutorial is a continuation of tutorials from Joy, so you should basically extend the document that is there in other PRs. Lemme know if you need more clear instruction

CynthiaPeter commented:
> lgmt so far, great and precise job 👏🏼
> 
> sorry for keeping you waiting 😞
> 
> just remember that when you will be adding examples of bindings for each section, this tutorial is a continuation of tutorials from Joy, so you should basically extend the document that is there in other PRs. Lemme know if you need more clear instruction

Yes. I understand that after Joy is done, Arya needs to edit the code before I use the code in my project.

CynthiaPeter commented:
Thank you, @derberg, for taking the time to review this draft.

For the next steps:
- I will use the code snippets from Joy -> Arya to complete my work and contact you for the code review and final feedback.

CynthiaPeter committed:
Updated changes

CynthiaPeter committed:
Removed index and section md

CynthiaPeter committed:
added code snippets from previous tutorial step

CynthiaPeter commented:
Hello @alequetzalli, @derberg, and @M3lkior,

I hope this meets you well. I just received the code snippets from [Arya's tutorial](https://github.com/asyncapi/website/pull/2331) and will need your feedback. I also have some questions below:

1. I currently just made edits for each binding snippet. I initially had the whole code for each section but it made the page too long. So, I thought that since it is a tutorial that's continued, I'd just use each server snippet. what do you think?

2. Based on the answer to the previous question, do I also combine all code snippets at the end of the tutorial and do I need to test it?

3. Arya's tutorial covered operations bindings, do I go ahead and add that to the tutorial too or not?
4. For the schemeregistryURL field under server binding, am I correct to add the $Ref link defined in Arya's tutorial?
5. How do I name the Kafka topic? Does it matter how it is named or it's just like a variable name? The question is regarding the topic field under channel bindings.

General feedback will be appreciated too. Thank you.

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+---
+title: Kafka-specific Bindings
```
```suggestion
title: Kafka specific bindings
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+---
+title: Kafka-specific Bindings
+description: In this tutorial, you'll learn how how to add specific Kafka information to an AsyncAPI document using bindings.
```
```suggestion
description: Learn how to configure Kafka bindings in your AsyncAPI document.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Introduction
+
+In a previous tutorial, you learnt how to manage schemas with schema registry. In this tutorial, you will learn how bindings with Kafka work by defining Kafka message key content and extending your AsyncAPI document with protocol-specific information.
```
```suggestion
In a previous tutorial, you learned how to manage schemas with a schema registry. This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+In a previous tutorial, you learnt how to manage schemas with schema registry. In this tutorial, you will learn how bindings with Kafka work by defining Kafka message key content and extending your AsyncAPI document with protocol-specific information.
+
+## Background Context
```
```suggestion
## Background context
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Background Context
+
+Bindings are crucial to an Event-driven application as they provide protocol-specific information that abstracts the details of how protocols handle messages from the core logic of your application. Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes ...
```
```suggestion
Bindings are essential for event-driven applications because they provide protocol-specific details, abstracting the complexities of message handling from your application's core logic. They enhance the API's clarity and usability by offering setup options and context for different protocols. Bindings include details like the topics your application reads from or writes to, message formatting, and rules for interacting with multiple data or messages.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+Bindings are crucial to an Event-driven application as they provide protocol-specific information that abstracts the details of how protocols handle messages from the core logic of your application. Bindings improve the understanding and usability of the API by offering more setup options and context for various protocols. It includes information such as topics an application reads from or writes ...
+
+In an AsyncAPI document, bindings can be appended to different document sections, such as servers, channels, or messages. They include standard information that is unique to a particular protocol. Bindings definitions allow you to specify the protocol-specific functionalities not supported by AsyncAPI's core features.
```
```suggestion
In an AsyncAPI document, bindings can be added to various sections like servers, channels, or messages. They contain protocol-specific details unique to each protocol. Binding definitions let you specify functionalities specific to the protocol, which are not covered by AsyncAPI's core features.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+In an AsyncAPI document, bindings can be appended to different document sections, such as servers, channels, or messages. They include standard information that is unique to a particular protocol. Bindings definitions allow you to specify the protocol-specific functionalities not supported by AsyncAPI's core features.
+
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on three levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object) and [message bindings](https://github.com/asyncapi/bindings/tr ...
```
```suggestion
You can configure several objects using Kafka bindings. However, for the scope of this tutorial, you will focus on three levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object) and [message bindings](https://github.com/asyncapi/bindings/tree/master/kafka#message-binding-object).
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, we are focusing on three levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object) and [message bindings](https://github.com/asyncapi/bindings/tr ...
+
+Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](pages/docs/tutorials/kafka/managing-schemas-using-schema-registry.md), you will add configurations for server, channel, and message bindings.
```
```suggestion
Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](pages/docs/tutorials/kafka/managing-schemas-using-schema-registry), you will add configurations for server, channel, and message bindings.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Add server bindings
+
+Server bindings provide protocol-specific configuration details for connecting and interacting with a server. For server bindings, we will add three fields to configure the server bindings. 
```
```suggestion
Server bindings provide protocol-specific configuration details for connecting and interacting with a server. For server bindings, we will add three fields: 
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Add channel bindings
+
+Channel bindings provide protocol-specific information for a particular channel. For example, in Kafka, you can specify number of partitions or replicas for a given topic.
```
```suggestion
Channel bindings provide protocol-specific information for a particular channel. In Kafka, you can specify a given topic's number of partitions or replicas.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Add message bindings
+
+Message bindings provide protocol-specific information for a specific message. Like the Kafka topic it's associated with, message key (if any), and Kafka-specific properties. 
```
```suggestion
Message bindings provide protocol-specific information for a specific message. Like the Kafka topic, it's associated with a message key (if any) and Kafka-specific properties. 
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+
+## Summary
+
+In this tutorial, you learned how to configure server and channel bindings. You also learned that bindings are essential in integrating Kafka with different systems, platforms, or protocols, especially in API specifications like AsyncAPI. 
```
```suggestion
In this tutorial, you learned how to configure server and channel bindings. You also learned that bindings are essential when integrating Kafka with different systems, platforms, or protocols — especially in API specifications like AsyncAPI. 
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+In this tutorial, you learned how to configure server and channel bindings. You also learned that bindings are essential in integrating Kafka with different systems, platforms, or protocols, especially in API specifications like AsyncAPI. 
+
+
+## Next Steps
```
```suggestion
## Next steps
```

Commit:
editorial rewrite

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+---
+title: Kafka specific bindings
```
```suggestion
title: Kafka bindings
```

Commit:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,95 @@
+title: Kafka bindings
+description: Learn how to configure Kafka bindings in your AsyncAPI document.
+---
+
```
@CynthiaPeter The only thing missing is the `page weight` metadata. Please add one, remembering that the tutorial order places THIS as the fourth and final Kafka tutorial. 😸

CynthiaPeter replied:
Done. Thanks

quetzalliwrites commented:
> 1. I currently just made edits for each binding snippet. I initially had the whole code for each section but it made the page too long. So, I thought that since it is a tutorial that's continued, I'd just use each server snippet. what do you think?
> 2. Based on the answer to the previous question, do I also combine all code snippets at the end of the tutorial and do I need to test it?
> 3. Arya's tutorial covered operations bindings, do I go ahead and add that to the tutorial too or not?
> 4. For the schemeregistryURL field under server binding, am I correct to add the $Ref link defined in Arya's tutorial?


I would like to hear @M3lkior opinions on **questions 1-4** since they are more about the technical content. @M3lkior what are your thoughts? 

Meanwhile, I have rewritten and cleaned up the writing, so I don't need to review this again until you make changes again or address any further feedback. 

> 5. How do I name the Kafka topic? Does it matter how it is named or it's just like a variable name? The question is regarding the topic field under channel bindings.

I think any sample name is fine since this is a sample application. Something descriptive and concise.

Thank you for your work on this @CynthiaPeter ❤️

quetzalliwrites committed:
Merge branch 'master' into feature/tutorial-bindings-with-kafka

CynthiaPeter committed:
Added metadata for page weight

CynthiaPeter committed:
Merge branch 'asyncapi:master' into feature/tutorial-bindings-with-kafka

M3lkior commented:
> > 1. I currently just made edits for each binding snippet. I initially had the whole code for each section but it made the page too long. So, I thought that since it is a tutorial that's continued, I'd just use each server snippet. what do you think?
> > 2. Based on the answer to the previous question, do I also combine all code snippets at the end of the tutorial and do I need to test it?
> > 3. Arya's tutorial covered operations bindings, do I go ahead and add that to the tutorial too or not?
> > 4. For the schemeregistryURL field under server binding, am I correct to add the $Ref link defined in Arya's tutorial?
> 
> I would like to hear @M3lkior opinions on **questions 1-4** since they are more about the technical content. @M3lkior what are your thoughts?
> 
> Meanwhile, I have rewritten and cleaned up the writing, so I don't need to review this again until you make changes again or address any further feedback.
> 
> > 5. How do I name the Kafka topic? Does it matter how it is named or it's just like a variable name? The question is regarding the topic field under channel bindings.
> 
> I think any sample name is fine since this is a sample application. Something descriptive and concise.
> 
> Thank you for your work on this @CynthiaPeter ❤️

Hey ; 

I only see this part in the Arya's tutorial : 

```
servers:
  kafkaServer:
    host: test.mykafkacluster.org:8092
    description: Kafka Server
    protocol: kafka
```
That not use bindings ;

Furthermore, a schemaRegistry url can be different from a kafka server url (if the server come from Confluent and the schema registry from Apicurio for example)

So for this tutorial, i think this is better to show how to define a schemaRegistryUrl property from scratch, no ?
(but the $ref is fine for other components that are not related to bindings from my side)

CynthiaPeter commented:
> > > 1. I currently just made edits for each binding snippet. I initially had the whole code for each section but it made the page too long. So, I thought that since it is a tutorial that's continued, I'd just use each server snippet. what do you think?
> > > 2. Based on the answer to the previous question, do I also combine all code snippets at the end of the tutorial and do I need to test it?
> > > 3. Arya's tutorial covered operations bindings, do I go ahead and add that to the tutorial too or not?
> > > 4. For the schemeregistryURL field under server binding, am I correct to add the $Ref link defined in Arya's tutorial?
> > 
> > 
> > I would like to hear @M3lkior opinions on **questions 1-4** since they are more about the technical content. @M3lkior what are your thoughts?
> > Meanwhile, I have rewritten and cleaned up the writing, so I don't need to review this again until you make changes again or address any further feedback.
> > > 5. How do I name the Kafka topic? Does it matter how it is named or it's just like a variable name? The question is regarding the topic field under channel bindings.
> > 
> > 
> > I think any sample name is fine since this is a sample application. Something descriptive and concise.
> > Thank you for your work on this @CynthiaPeter ❤️
> 
> Hey ;
> 
> I only see this part in the Arya's tutorial :
> 
> ```
> servers:
>   kafkaServer:
>     host: test.mykafkacluster.org:8092
>     description: Kafka Server
>     protocol: kafka
> ```
> 
> That not use bindings ;
> 
> Furthermore, a schemaRegistry url can be different from a kafka server url (if the server come from Confluent and the schema registry from Apicurio for example)
> 
> So for this tutorial, i think this is better to show how to define a schemaRegistryUrl property from scratch, no ? (but the $ref is fine for other components that are not related to bindings from my side)

Thank you, @M3lkior 
I am lost. Do you have time to chat anytime soon?

M3lkior commented:
> > > > 1. I currently just made edits for each binding snippet. I initially had the whole code for each section but it made the page too long. So, I thought that since it is a tutorial that's continued, I'd just use each server snippet. what do you think?
> > > > 2. Based on the answer to the previous question, do I also combine all code snippets at the end of the tutorial and do I need to test it?
> > > > 3. Arya's tutorial covered operations bindings, do I go ahead and add that to the tutorial too or not?
> > > > 4. For the schemeregistryURL field under server binding, am I correct to add the $Ref link defined in Arya's tutorial?
> > > 
> > > 
> > > I would like to hear @M3lkior opinions on **questions 1-4** since they are more about the technical content. @M3lkior what are your thoughts?
> > > Meanwhile, I have rewritten and cleaned up the writing, so I don't need to review this again until you make changes again or address any further feedback.
> > > > 5. How do I name the Kafka topic? Does it matter how it is named or it's just like a variable name? The question is regarding the topic field under channel bindings.
> > > 
> > > 
> > > I think any sample name is fine since this is a sample application. Something descriptive and concise.
> > > Thank you for your work on this @CynthiaPeter ❤️
> > 
> > 
> > Hey ;
> > I only see this part in the Arya's tutorial :
> > ```
> > servers:
> >   kafkaServer:
> >     host: test.mykafkacluster.org:8092
> >     description: Kafka Server
> >     protocol: kafka
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That not use bindings ;
> > Furthermore, a schemaRegistry url can be different from a kafka server url (if the server come from Confluent and the schema registry from Apicurio for example)
> > So for this tutorial, i think this is better to show how to define a schemaRegistryUrl property from scratch, no ? (but the $ref is fine for other components that are not related to bindings from my side)
> 
> Thank you, @M3lkior I am lost. Do you have time to chat anytime soon?

Not this week unfortunatly, i am so busy ; but can you send me a draft of your proposal ? Maybe i don't understand what you want to do with the $ref

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,96 @@
+  production:
+    bindings:
+      kafka:
+        schemaRegistryUrl: '$ref: http://localhost:8080/apis/registry/v2/groups/my-group/artifacts/UserSignedUp'
```
the url must be the base path of the registry api 
the `$ref` must referencing a component or an external file, so this is not appropriate here
```suggestion
        schemaRegistryUrl: 'http://localhost:8080/apis/registry/'
```

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

the url must be the base path of the registry api
the $ref must referencing a component or an external file, so this is not appropriate here.

Co-authored-by: Ludovic Dussart <ludovic.dussart.pro@gmail.com>

derberg reviewed (commented):
lgtm, the only 2 things missing are:

- imho at the end of tutorial we should always have a final, full AsyncAPI document that is easy to copy/paste in studio for example. Look at https://www.asyncapi.com/docs/tutorials/websocket#define-operations as example
- we have nice examples of bindings but without any explanation:
   - why do you add `bindingVersion` everywhere? what does it mean? what if you do not put it in?
   - why in channel binding you have `kafka.topic` set to `UserSignedUp` but the `channel.address` is `user_signedup`. How does AsyncAPI channel address relates to kafka binding and `topic` field?
   - for some fields you have explanation already in bindings like https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object

CynthiaPeter commented:
> lgtm, the only 2 things missing are:
> 
> * imho at the end of tutorial we should always have a final, full AsyncAPI document that is easy to copy/paste in studio for example. Look at https://www.asyncapi.com/docs/tutorials/websocket#define-operations as example
> * we have nice examples of bindings but without any explanation:
>   
>   * why do you add `bindingVersion` everywhere? what does it mean? what if you do not put it in?
>   * why in channel binding you have `kafka.topic` set to `UserSignedUp` but the `channel.address` is `user_signedup`. How does AsyncAPI channel address relates to kafka binding and `topic` field?
>   * for some fields you have explanation already in bindings like https://github.com/asyncapi/bindings/tree/master/kafka#channel-binding-object

Thank you @derberg for the feedback. 
I'll implement this feedback and share an update with you, ASAP.

M3lkior commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,96 @@
+        $ref: '#/components/messages/userSignedUp'
+    bindings:
+      kafka:
+        topic: 'UserSignedUp'
```
we needs to maybe set a more different name to illustrate this documentation 

```
Kafka topic name if different from channel name.
```

CynthiaPeter replied:
@M3lkior , I read that `topic` is to Kafka what `channel` is to AsyncAPI. If that is the case, how is Kafka topic name different from channel name?

and how is that different from `channel.address`?

lbroudoux replied:
`topic` was introduced in AsyncAPI v2 as channel was much more a "logical view", and topic represents a living object in Kafka broker. Moreover, a Kafka topic has naming conventions that were different from a channel key in AsyncAPI v2.

In AsyncAPI v3, we now have `channel.address` and I think this drops the need for `topic` as we'll probably have `channel.address === kafka.topic` most of the time - as the representation of the technical object living in a Kafka broker. That said, we absolutely need to check the naming conventions of both properties to be sure they are a perfect match.

Generally speaking - getting beyond Kafka here - I'm not sure `channel.address` will be generic and future-proof enough to handle all the protocol naming specificities (think of `/`, `-`, `_`, `.` chars that may or may not accepted depending on protocol...). I think it's nice to let the binding override the `channel.address` property if needed.

Coming back to the initial question, I like @M3lkior suggestion. I would put something like this in the doc:

```
Kafka topic name if different from channel name - if you need to follow specific naming conventions for exampee.
```

derberg replied:
@CynthiaPeter from the slack discussion I figure, the best is just to remove `topic` from this tutorial.

and I guess folks we pinged (kafka binding maintainers), if there are doubts and clarifications needed, they will take it from here and followup in `bindings` repo

CynthiaPeter committed:
Updated document based on feedback on PR

CynthiaPeter commented:
Thank you @derberg and @M3lkior for the feedback.

Based on the feedback you shared, I have made the following improvements:
- At the end of the tutorial, I combined all the bindings to make one AsyncAPI document. However, I pasted it in the Studio tool and ran into some bugs and might need help with debugging or figuring things out. Especially around channel bindings.

- I removed the `bindingversion` field and placed it in just one binding. At first it was in every binding code snippet because it is the one value that cuts across all 4 bindings. However, it is also optional, so I added a note to indicate this information for the reader.

**Some questions:**

- I read that `topic` is to Kafka what `channel` is to AsyncAPI. If that is the case, how is Kafka topic name different from channel name? and how is that different from channel.address field?

CynthiaPeter committed:
updated based on feedback from Dale. Used just channel.address

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+      kafka:
+        schemaRegistryUrl: 'http://localhost:8080/apis/registry/'
+        schemaRegistryVendor: 'apicurio'
+        bindingVersion: '0.5.0'
```
please always validate with studio if your snippets are correct

```suggestion
servers:
  kafkaServer:
    host: test.mykafkacluster.org:8092
    description: Kafka Server
    protocol: kafka
    bindings:
      kafka:
        schemaRegistryUrl: 'http://localhost:8080/apis/registry/'
        schemaRegistryVendor: 'apicurio'
        bindingVersion: '0.4.0'
```

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
````diff
@@ -0,0 +1,183 @@
+
+Congratulations, you've completed the tutorial! Putting these blocks together gives you an AsyncAPI document all ready to go.
+
+```yaml
````
please adjust it to reflect final version containing all that was added above + things from previous example

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+
+## Background context
+
+Bindings are essential for event-driven applications because they provide protocol-specific details, abstracting the complexities of message handling from your application's core logic. They enhance the API's clarity and usability by offering setup options and context for different protocols. Bindings include details like the topics your application reads from or writes to, message formatting, an ...
```
please also link to https://www.asyncapi.com/docs/concepts/asyncapi-document/adding-bindings

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, you will focus on four levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [operations binding](https://github.com/asyncapi/bindings/tree/master/kafka#operation-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tre ...
+
+Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](pages/docs/tutorials/kafka/managing-schemas-using-schema-registry), you will add configurations for server, operations, channel, and message bindings.
```
```suggestion
Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](/docs/tutorials/kafka/managing-schemas-using-schema-registry), you will add configurations for server, operations, channel, and message bindings.
```

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, you will focus on four levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [operations binding](https://github.com/asyncapi/bindings/tree/master/kafka#operation-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tre ...
+
+Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](pages/docs/tutorials/kafka/managing-schemas-using-schema-registry), you will add configurations for server, operations, channel, and message bindings.
+
```
you folks need to be consistent across these tutorials, so they are easy to follow 1-4, but also are useful for someone that just wants to do tut 4

your tutorial is a continuation after tutorial 3, look at it https://github.com/asyncapi/website/pull/2331/files#diff-db9dfd2002015a20cf7de2d3c2ce8e68a1537fde911f556b307e238fc3e1f09bR18. It says that it is next in the line, but also to make it easier for the reader, it shows the final version of the document that was created in tutorial 2. It shows what document will be modified, and then at the end shows modified version

your tutorial shows only final version after you finish tutorial. Would be great to show initial version of the document that you later extend with bindings

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+          enum: ['myGroupId']
+        clientId:
+          type: string
+          enum: ['myClientId']
```
```suggestion
operations:
  onUserSignedUp:
    action: receive
    channel:
      $ref: '#/channels/userSignedUp'
    bindings:
      kafka:
        bindingVersion: '0.4.0'
        groupId:
          type: string
          enum: ['myGroupId']
        clientId:
          type: string
          enum: ['myClientId']
```

CynthiaPeter replied:
Is there a particular reason why we have bindingVersion specific to every configuration? Also, does this mean that in the final version, we have 4/5 different bindingVersion or we have just one?

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+          retention.ms: 604800000
+          retention.bytes: 1000000000
+          delete.retention.ms: 86400000
+          max.message.bytes: 1048588
```
```suggestion
      kafka:
        bindingVersion: '0.4.0'
        topic: 'UserSignedUp'
        partitions: 10
        replicas: 2
        topicConfiguration:
          cleanup.policy: ["delete", "compact"]
          retention.ms: 604800000
          retention.bytes: 1000000000
          delete.retention.ms: 86400000
          max.message.bytes: 1048588
```

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+  messages:
+    userSignedUp:
+      bindings:
+        kafka:
+            key:
+              type: string
+              enum: ['myKey']
+            schemaIdLocation: 'payload'
+            schemaIdPayloadEncoding: 'apicurio-new'
+            schemaLookupStrategy: 'TopicIdStrategy'
+            bindingVersion: '0.5.0'
```
```suggestion
        kafka:
            key:
              type: string
              enum: ['myKey']
            schemaIdLocation: 'payload'
            schemaIdPayloadEncoding: 'apicurio-new'
            schemaLookupStrategy: 'TopicIdStrategy'
            bindingVersion: '0.4.0'
```

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+
+In an AsyncAPI document, bindings can be added to various sections like servers, channels, or messages. They contain protocol-specific details unique to each protocol. Binding definitions let you specify functionalities specific to the protocol, which are not covered by AsyncAPI's core features.
+
+You can configure several objects using Kafka bindings. However, for the scope of this tutorial, you will focus on four levels of bindings: [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object), [operations binding](https://github.com/asyncapi/bindings/tree/master/kafka#operation-binding-object), [channel bindings](https://github.com/asyncapi/bindings/tre ...
```
please consider modification of this paragraph

here you should just have an intro to kafka binding and point to https://github.com/asyncapi/bindings/tree/master/kafka

then the links that you have, to specific objects, like message or others, they are useful, but super useful they would be if instead of here, they are in respective sections. So for example, as reader I would prefer to have a link to server kafka binding in server-related section, like a sentence below each code snippet

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+
+## Add server bindings
+
+Server bindings provide protocol-specific configuration details for connecting and interacting with a server. For server bindings, we will add three fields: 
```
I think I left that comment some time ago already

we need intros to these snippets

in this one for example you add schema registry details, why?

for example here you explain: server binding for example allows you to specify for a given server, where is related schema registry located. Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry read https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation

you know what I mean.

you do not have to go through each property, they are well explained in bindings https://github.com/asyncapi/bindings/tree/master/kafka

but an intro makes sense - so with example they understand the concept behind it

VaishnaviNandakumar commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,183 @@
+
+## Introduction
+
+In a previous tutorial, you learned how to manage schemas with a schema registry. This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```
@CynthiaPeter could you link the previous tutorial here?

CynthiaPeter replied:
Thank you, Vaishnavi. Done.

CynthiaPeter committed:
update docs based on feedback

CynthiaPeter committed:
removed production

CynthiaPeter committed:
edited bindingVersion

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+
+Server bindings allows you to specify a `schemaRegistryUrl` which provides an API Url for a given server where a schema registry was used. A Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry read [this guide on message validation](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
+
+`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case "Apicurio".
```
```suggestion
`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case Apicurio Registry.
```

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+
+[Learn more about server specific bindings.](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
+
+> Important: `BindingVersion` is the field version of a binding. It specifies the version of the binding specification that is used to describe how an API interacts with Kafka. The `bindingVersion` field is an optional field that is available for all bindings and ensures that as the specifications for different messaging systems evolve, APIs based on these specifications can remain consistent, co ...
```
```suggestion
> Important: `bindingVersion` is the field version of a binding. It specifies the version of the binding specification that is used to describe how an API interacts with Kafka. The `bindingVersion` field is an optional field that is available for all bindings and ensures that as the specifications for different messaging systems evolve, APIs based on these specifications can remain consistent, compatible, and maintainable over time.  
```

if we refer to key name, we always have to keep original capitalization to be technically accura

derberg replied:
sorry but I have no idea what `The bindingVersion field is an optional field that is available for all bindings and ensures that as the specifications for different messaging systems evolve, APIs based on these specifications can remain consistent, compatible, and maintainable over time.` means. Let's not overcomplicate and add sugar in tech docs.

 ```suggestion
> Important: Always utilize the `bindingVersion` field in your production environment. This field allows you to specify the exact version of the binding you're using. Failure to include this field may result in unintended usage of the latest version, potentially causing issues due to possible future breaking changes.
```

CynthiaPeter replied:
@derberg The intention was not to add sugar. I understood that the binding version was important for each bindings because of compatibility. And knowing the bindingVersion for each bindings ensures consistency as well. 

However, I have removed the sentence.

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
````diff
@@ -0,0 +1,242 @@
+        bindingVersion: '0.4.0'
+```
+
+[Learn more about server specific bindings.](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
````
please always look at your final document in a preview and read it

it is important to be consistent, 

here you put reference to more info about bindings after a code snippet, as a full text link - then in other sections you do it completely different, and before the code snippet

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+      payload:
+        schemaFormat: 'application/vnd.apache.avro+json;version=1.9.0'
+        schema:
+          $ref: http://localhost:8080/apis/registry/v2/groups/my-group/artifacts/UserSignedUp          
```
```suggestion
asyncapi: 3.0.0
info:
  title: User Signup API
  version: 1.0.0
  description: The API notifies you whenever a new user signs up in the application.
servers:
  kafkaServer:
    host: test.mykafkacluster.org:8092
    description: Kafka Server
    protocol: kafka
    bindings:
      kafka:
        schemaRegistryUrl: 'http://localhost:8080/apis/registry/'
        schemaRegistryVendor: 'apicurio'
        bindingVersion: '0.4.0'
operations:
  onUserSignedUp:
    action: receive
    channel:
      $ref: '#/channels/userSignedUp'
    bindings:
      kafka:
        bindingVersion: '0.4.0'
        groupId:
          type: string
          enum: ['myGroupId']
        clientId:
          type: string
          enum: ['myClientId']
channels:
  userSignedUp:
    description: This channel contains a message per each user who signs up in our application.
    address: user_signedup
    messages:
      userSignedUp:
        $ref: '#/components/messages/userSignedUp'
    bindings:
      kafka:
        bindingVersion: '0.4.0'
        partitions: 10
        replicas: 2
        topicConfiguration:
          cleanup.policy: ["delete", "compact"]
          retention.ms: 604800000
          retention.bytes: 1000000000
          delete.retention.ms: 86400000
          max.message.bytes: 1048588
components:
  messages:
    userSignedUp:
      bindings:
        kafka:
          bindingVersion: '0.4.0'
          key:
            type: string
            enum: ['myKey']
          schemaIdLocation: 'payload'
          schemaIdPayloadEncoding: 'apicurio-new'
          schemaLookupStrategy: 'TopicIdStrategy'
      payload:
        schemaFormat: 'application/vnd.apache.avro+json;version=1.9.0'
        schema:
          $ref: http://localhost:8080/apis/registry/v2/groups/my-group/artifacts/UserSignedUp          
```

derberg replied:
as I wrote in previous comments, you always need o make sure example you share and promote is actually valid

dalelane commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+      kafka:
+        schemaRegistryUrl: 'http://localhost:8080/apis/registry/'
+        schemaRegistryVendor: 'apicurio'
+        bindingVersion: '0.4.0'
```
```suggestion
        bindingVersion: '0.5.0'
```

CynthiaPeter replied:
Hello @dalelane, I had a chat with @derberg about bindingVersion and he explained that we use the compatible version to studio in our docs because we have a bug in validation and studio complains about 0.5.0. 

Is this ok?

derberg replied:
@CynthiaPeter not anymore, sorry, wrote you in slack, `0.5.0` is definitely fine now

CynthiaPeter replied:
I understand. Thank you.

dalelane commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+      $ref: '#/channels/userSignedUp'
+    bindings:
+      kafka:
+        bindingVersion: '0.4.0'
```
```suggestion
        bindingVersion: '0.5.0'
```

dalelane commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+
+Channel bindings provide protocol-specific information for a particular channel.
+
+These configurations may include information about how to connect with the Kafka topic, plus any specific settings or behaviors to be used. The Channel Binding Object is part of AsyncAPI's wider bindings architecture, which specifies how the API interacts with the messaging system — in this case, Kafka.
```
```suggestion
These configurations may include information how the Kafka topic has been configured. The Channel Binding Object is part of AsyncAPI's wider bindings architecture, which specifies how the API interacts with the messaging system — in this case, Kafka.
```

I bumped a little on saying the Kafka channel bindings are about how to connect, as they are more intended to describe the topic config. I do accept that you could describe deciding the number of consumers to run in a group as being information about how to connect, but that feels like a bit of a stretch IMO.

dalelane commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+
+## Add message bindings
+
+Message bindings provide protocol-specific information for a specific message. Like the Kafka topic, it's associated with a message key (if any) and Kafka-specific properties. 
```
```suggestion
Message bindings provide protocol-specific information for a specific message. For Kafka topics, this can include how message keys are used, and details about how serialized message data has been encoded. 
```

dalelane commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,242 @@
+            schemaIdLocation: 'payload'
+            schemaIdPayloadEncoding: 'apicurio-new'
+            schemaLookupStrategy: 'TopicIdStrategy'
+            bindingVersion: '0.4.0'
```
```suggestion
            bindingVersion: '0.5.0'
```

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Dale Lane <dale.lane@gmail.com>

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Dale Lane <dale.lane@gmail.com>

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Valid document

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

CynthiaPeter committed:
update docs

CynthiaPeter committed:
fixed updates

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Dale Lane <dale.lane@gmail.com>

CynthiaPeter committed:
Update pages/docs/tutorials/kafka/bindings-with-kafka.md

Co-authored-by: Dale Lane <dale.lane@gmail.com>

CynthiaPeter committed:
update docs

CynthiaPeter committed:
Merge branch 'feature/tutorial-bindings-with-kafka' of https://github.com/CynthiaPeter/website into feature/tutorial-bindings-with-kafka

CynthiaPeter committed:
removed the note on kafka.topic

CynthiaPeter committed:
fixed link error

derberg commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+## Introduction
+
+In a [previous tutorial](/docs/tutorials/kafka/managing-schemas-using-schema-registry.md), you learned how to manage schemas with a schema registry. This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```
```suggestion
In a [previous tutorial](/docs/tutorials/kafka/managing-schemas-using-schema-registry), you learned how to manage schemas with a schema registry. This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```

derberg reviewed (dismissed):
Thanks so much @CynthiaPeter 
LGTM from my side big time!!!

@alequetzalli on to you, I guess you do not want to merge yet as it is mentioning in intro tutorial that is not yet published

derberg committed:
Merge branch 'master' into feature/tutorial-bindings-with-kafka

dalelane reviewed (approved):
thanks for doing this - I think this will be useful

quetzalliwrites committed:
Merge branch 'master' into feature/tutorial-bindings-with-kafka

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+In an AsyncAPI document, bindings can be added to various sections like servers, channels, or messages. They contain protocol-specific details unique to each protocol. Binding definitions let you specify functionalities specific to the protocol, which are not covered by AsyncAPI's core features.
+
+You can configure several objects using [Kafka bindings](https://github.com/asyncapi/bindings/tree/master/kafka#readme). However, for the scope of this tutorial, you will focus on four levels of bindings: server bindings, operations binding, channel bindings and message bindings.
```
```suggestion
You can configure several objects using [Kafka bindings](https://github.com/asyncapi/bindings/tree/master/kafka#readme). However, for the scope of this tutorial, you will focus on four levels of bindings: server bindings, operations binding, channel bindings, and message bindings.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+Server bindings provide protocol-specific configuration details for connecting and interacting with a server. 
+
+Server bindings allows you to specify a `schemaRegistryUrl` which provides an API Url for a given server where a schema registry was used. A Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry read [this guide on message validation](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
```
```suggestion
Server bindings allow you to specify a `schemaRegistryUrl`, which provides an API URL for a given server where a schema registry was used. A Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry, read the [message validation guide for schema registry](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+Server bindings allows you to specify a `schemaRegistryUrl` which provides an API Url for a given server where a schema registry was used. A Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry read [this guide on message validation](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
+
+`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case Apicurio Registry. Learn more about other fields you can configure under [server bindings.](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
```
```suggestion
`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case, Apicurio Registry. Learn about other fields you can configure under [server bindings.](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
```

quetzalliwrites reviewed (dismissed):
AMAZING WORK, @CynthiaPeter. Thank you ✨ 😻

Commit:
grammar fixes

quetzalliwrites reviewed (dismissed):
:✅

quetzalliwrites commented:
> @alequetzalli on to you, I guess you do not want to merge yet as it is mentioning in intro tutorial that is not yet published

Correct @derberg, we first gotta merge @Arya-Gupta's tutorial.

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+Using the code snippets from the previous tutorial, where you learned [how to manage Avro schemas using a centralized schema registry that enables you to share schemas across multiple applications](/docs/tutorials/kafka/managing-schemas-using-schema-registry), you will add configurations for server, operations, channel, and message bindings.
+
+Below, you can find the updated Schema reference file that we'll be using for this tutorial.
```
```suggestion
Below, you can find the updated schema reference file you'll use for this tutorial.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+Server bindings provide protocol-specific configuration details for connecting and interacting with a server. 
+
+Server bindings allow you to specify a `schemaRegistryUrl`, which provides an API URL for a given server where a schema registry was used. A Schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry, read the [message validation guide for schema registry](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
```
```suggestion
Server bindings allow you to specify a `schemaRegistryUrl`, which provides an API URL for a given server where a schema registry was used. A schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry, read the [message validation guide for schema registry](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
```

Commit:
grammar fixes

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+## Introduction
+
+In a [previous tutorial](/docs/tutorials/kafka/managing-schemas-using-schema-registry), you learned how to manage schemas with a schema registry. This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```
```suggestion
[You learned how to manage schemas with a schema registry in the previous tutorial](/docs/tutorials/kafka/managing-schemas-using-schema-registry). This tutorial teaches you how Kafka bindings function by defining Kafka messages and expanding your AsyncAPI document with protocol-specific details.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+## Background context
+
+Bindings are essential for event-driven applications because they provide protocol-specific details, abstracting the complexities of message handling from your application's core logic. They enhance the API's clarity and usability by offering setup options and context for different protocols. [Bindings](docs/concepts/asyncapi-document/adding-bindings) include details like the topics your applicat ...
```
```suggestion
Bindings are essential for event-driven applications because they provide protocol-specific details, abstracting the complexities of message handling from your application's core logic. They enhance the API's clarity and usability by offering setup options and context for different protocols. Bindings include the topics your application reads from or writes to, message formatting, and rules for interacting with multiple data or messages.
```

quetzalliwrites commented on pages/docs/tutorials/kafka/bindings-with-kafka.md:
```diff
@@ -0,0 +1,236 @@
+
+Server bindings allow you to specify a `schemaRegistryUrl`, which provides an API URL for a given server where a schema registry was used. A schema registry is a repository for managing and validating messages' schemas. To learn more about schema registry, read the [message validation guide for schema registry](https://www.asyncapi.com/docs/guides/message-validation#schema-registry-validation).
+
+`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case, Apicurio Registry. Learn about other fields you can configure under [server bindings.](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
```
```suggestion
`schemaRegistryVendor` is used optionally to refer to vendors or platforms that provide the schema registry service, in this case, Apicurio Registry. Learn about other fields you can configure under [server bindings](https://github.com/asyncapi/bindings/tree/master/kafka#server-binding-object).
```

Commit:
fix broken links

quetzalliwrites committed:
Merge branch 'master' into feature/tutorial-bindings-with-kafka

quetzalliwrites commented:
/rtm
